cdap-app-fabric/src/main/java/co/cask/cdap/extension/ExtensionLoader.java

```diff
@@ -0,0 +1,227 @@
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A class to maintain a cache of extensions available in a configured extensions directory. It uses the Java
+ * {@link ServiceLoader} architecture to load extensions from the specified directory. It first tries to load
+ * extensions using a {@link URLClassLoader} consisting of all jars in the configured extensions directory. For unit
+ * tests, it loads extensions using the system classloader.
+ *
+ * The supported directory structure is:
+ * <pre>
+ *   [extensions-directory]/[extensions-module-directory]/
+ *   [extensions-directory]/[extensions-module-directory]/file-containing-extensions-1.jar
+ *   [extensions-directory]/[extensions-module-directory]/file-containing-extensions-2.jar
+ *   ...
+ *   [extensions-directory]/[extensions-module-directory]/file-containing-extensions-n.jar
+ * </pre>
+ *
+ * Each extensions jar file above can contain multiple extensions.
+ * 
+ * @param <CACHE_KEY> the key type stored in the cache
+ * @param <CACHE_VALUE> the value type stored in the cache
+ */
+public class ExtensionLoader<CACHE_KEY, CACHE_VALUE> {
+  private static final Logger LOG = LoggerFactory.getLogger(ExtensionLoader.class);
+
+  private final List<String> extDirs;
+  private final Class<CACHE_VALUE> cacheValueClass;
+  // The ServiceLoader that loads extension implementation from the CDAP system classloader.
+  private final ServiceLoader<CACHE_VALUE> systemExtensionLoader;
+  private final Function<CACHE_VALUE, Set<CACHE_KEY>> extensionToKeys;
+  private final LoadingCache<CACHE_KEY, CACHE_VALUE> extensionsCache;
+  private final CACHE_VALUE defaultExtension;
+
+  public ExtensionLoader(String extDirs, Function<CACHE_VALUE, Set<CACHE_KEY>> extensionToKeys,
+                         Class<CACHE_VALUE> cacheValueClass, CACHE_VALUE defaultExtension) {
+    this.extDirs = ImmutableList.copyOf(Splitter.on(';').omitEmptyStrings().trimResults().split(extDirs));
+    this.cacheValueClass = cacheValueClass;
+    this.systemExtensionLoader = ServiceLoader.load(cacheValueClass);
+    this.extensionToKeys = extensionToKeys;
+    this.extensionsCache = createExtensionsCache();
+    this.defaultExtension = defaultExtension;
+  }
+
+  /**
+   * Returns the extension for the specified key.
+   */
+  public CACHE_VALUE getExtension(CACHE_KEY key) {
+    return extensionsCache.getUnchecked(key);
+  }
+
+  /**
+   * Returns all cached extensions.
+   */
+  public Map<CACHE_KEY, CACHE_VALUE> listExtensions() {
+    return extensionsCache.asMap();
+  }
+
+  /**
+   * Preloads the extension cache with all the extensions from the specified extensions directory.
+   */
+  public void preload() {
```

Comment:
I don't like this contract. Why there is preload? If the contract of this class is a extension loader, it shouldn't expose internal operations (e.g. that it is a cache).